File: AdventOfCode2020/Day4.cpp
#include <iostream>
#include <fstream>
#include <string>
using namespace std;
/*
iyr:2015 cid:189 ecl:oth byr:1947 hcl:#6c4ab1 eyr:2026
hgt:174cm
pid:526744288*/
int main()
{
	bool ecl, pid, eyr, hcl, byr, iyr, cid, hgt=false;
	ecl = pid = eyr = hcl = byr = iyr = cid = hgt = false;
	ifstream infile;
	string input;
	string temp;
	int counter = 0;
	infile.open("E:\\AdventOfCode\\AdventOfCode2020\\input1.txt");
	//temp.clear();
	//getline(infile, input);
	while (!infile.eof())
	{
		temp.clear();
		getline(infile, input);
		if (!input.size())
		{
			ecl = pid = eyr = hcl = byr = iyr = cid = hgt = false;
		}
		while (input.size())
		{

			temp.clear();
			if (input[0] == 'e' && input[1] == 'y' && input[2] == 'r')
			{
				temp.clear();
				eyr = true;
				input.erase(0,4);
				while (input[0] != ' '&& input.size())
				{
					temp.push_back(input[0]);
					input.erase(input.begin());
				}
				if (input.size())
				{
					input.erase(input.begin());
				}
				if (temp.size() != 4 || (stoi(temp) < 2020 || stoi(temp) > 2030))
				{
					eyr = false;
					//cout << stoi(temp);
				}
				else
				{
					//cout << temp << endl;
				}
				//cout << stoi(temp);
			}
			else if (input[0] == 'i' && input[1] == 'y' && input[2] == 'r')
			{
				temp.clear();
				iyr = true;
				input.erase(0, 4);
				while (input[0] != ' ' && input.size())
				{
					temp.push_back(input[0]);
					input.erase(input.begin());
				}
				if (input.size())
				{
					input.erase(input.begin());
				}
				if (temp.size() != 4 || (stoi(temp) < 2010 || stoi(temp) > 2020))
				{
					iyr = false;
					//cout << stoi(temp);
				}
				else
				{
					//cout << temp << endl;
				}
			}
			else if (input[0] == 'e' && input[1] == 'c' && input[2] == 'l')
			{

				temp.clear();

				if (ecl)
				{
					ecl = false;
				}
				else
				{


					ecl = true;
				}
				input.erase(0, 4);
				while (input[0] != ' ' && input.size())
				{
					temp.push_back(input[0]);
					input.erase(input.begin());
				}
				if (input.size())
				{
					input.erase(input.begin());
				}
				if (temp.size() != 3 || (temp != "amb" && temp != "blu" && temp != "brn" && temp != "gry" && temp != "grn" && temp != "hzl" && temp != "oth"))
				{
					ecl = false;
				}
				else
				{
					//cout << temp << endl;
				}
			}
			else if ((input[0] == 'p' && input[1] == 'i' && input[2] == 'd'))
			{
				temp.clear();
				pid = true;
				input.erase(0, 4);
				while (input[0] != ' ' && input.size())
				{
					temp.push_back(input[0]);
					input.erase(input.begin());
				}
				if (input.size())
				{
					input.erase(input.begin());
				}
				if (temp.size() != 9)
				{
					pid = false;
				}
				else
				{
					for (int i = 0; i < temp.size(); i++)
					{
						if (temp[i] >= '0' && temp[i] <= '9')
						{

						}
						else
						{

							pid = false;
						}
					}
				}
				if (pid)
				{
					//cout << temp << endl;
				}
			}
			else if ((input[0] == 'h' && input[1] == 'c' && input[2] == 'l'))
			{
				hcl = true;
				input.erase(0, 4);
				while (input[0] != ' ' && input.size())
				{
					temp.push_back(input[0]);
					input.erase(input.begin());
				}
				if (input.size())
				{
					input.erase(input.begin());
				}
				if (temp[0] != '#')
				{
					hcl = false;
					//cout << temp << endl;
				}
				else
				{
					temp.erase(temp.begin());
				}
				if (temp.size()!=6)
				{
					
					hcl = false;
				}
				else
				{
					for (int i = 0; i < temp.size(); i++)
					{
						if ((temp[i] >= '0' && temp[i] <='9') || (temp[i] >= 'a' && temp[i]<='f'))
						{
//							cout << "Within parameters" << endl;
						}
						else
						{
							hcl = false;
						}
					}
				}
				if (hcl)
				{
					//cout << temp << endl;
				}
			}
			else if ((input[0] == 'b' && input[1] == 'y' && input[2] == 'r'))
			{
				byr = true;
				input.erase(0, 4);
				while (input[0] != ' ' && input.size())
				{
					temp.push_back(input[0]);
					input.erase(input.begin());
				}
				if (input.size())
				{
					input.erase(input.begin());
				}
				if (temp.size() != 4 || (stoi(temp) < 1920 || stoi(temp) > 2002))
				{
					byr = false;
					//cout << stoi(temp);
				}
				//else
					//cout << temp << endl;
			}
			else if ((input[0] == 'c' && input[1] == 'i' && input[2] == 'd'))
			{
				cid = true;
				input.erase(0, 4);
				while (input[0] != ' ' && input.size())
				{
					temp.push_back(input[0]);
					input.erase(input.begin());
				}
				if (input.size())
				{
					input.erase(input.begin());
				}
			}
			else if ((input[0] == 'h' && input[1] == 'g' && input[2] == 't'))
			{
			//hgt:174cm
				hgt = true;
				input.erase(0, 4);
				while (input[0] != ' ' && input.size())
				{
					temp.push_back(input[0]);
					input.erase(input.begin());
				}
				if (input.size())
				{
					input.erase(input.begin());// to delete the space
				}
				if (temp[temp.size() - 1] == 'm' && temp[temp.size() - 2] == 'c')
				{
					temp.erase(temp.size() - 2, 2);
					if (stoi(temp) < 150 || stoi(temp) > 193)
					{
						hgt = false;
					}
				}
				else if (temp[temp.size() - 1] == 'n' && temp[temp.size() - 2] == 'i')
				{
					temp.erase(temp.size() - 2, 2);
					if (stoi(temp) < 59 || stoi(temp) > 76)
					{
						hgt = false;
					}
				}
				else
					hgt = false;
				if (hgt)
				{
					cout << temp << endl;
				}
			}
		}

			
			if (ecl && pid && eyr && hcl && byr && iyr && hgt)
			{
				counter++;
				ecl = pid = eyr = hcl = byr = iyr = cid = hgt = false;
				temp.clear();
				input.clear();
			}
			// dont run following loop

		//temp.clear();
		//getline(infile, input);
		
	}
	/*if (ecl && pid && eyr && hcl && byr && iyr && hgt)
	{
		counter++;
	}*/
	cout << counter << endl;
}

File: AdventOfCode2020/Day4.py
import re
import itertools
class Checker:
    '''
    byr=False
    iyr=False
    eyr=False
    hgt=False
    hcl=False
    ecl=False
    pid=False
    cid=False
    counter=0
    '''
    def __init__(self):
        self.byr=False
        self.iyr=False
        self.eyr=False
        self.hgt=False
        self.hcl=False
        self.ecl=False
        self.pid=False
        self.cid=False
        self.counter=0
    def reset(self):
        if(self.byr==True and self.iyr==True and self.eyr==True and self.hgt==True and self.hcl==True and self.ecl==True and
        self.pid==True):
            self.counter+=1
        self.byr=False
        self.iyr=False
        self.eyr=False
        self.hgt=False
        self.hcl=False
        self.ecl=False
        self.pid=False
        self.cid=False
    def check(self):
        if(self.byr==True and self.iyr==True and self.eyr==True and self.hgt==True and self.hcl==True and self.ecl==True and
        self.pid==True):
            self.counter+=1
            self.byr=False
            self.iyr=False
            self.eyr=False
            self.hgt=False
            self.hcl=False
            self.ecl=False
            self.pid=False
            self.cid=False
        
    def checker(self,statement):
        arr=statement.split(":")
        if arr==['']:
            self.reset()
        elif(arr[0]=="byr" and len(arr[1])==4 and int(arr[1])<=2002 and int(arr[1])>=1920):
            self.byr=True
            print(arr[1])
        elif arr[0]=="iyr" and len(arr[1])==4 and int(arr[1])<=2020 and int(arr[1])>=2010:
            self.iyr=True
        elif arr[0]=="eyr"and len(arr[1])==4 and int(arr[1])<=2030 and int(arr[1])>=2020:
            self.eyr=True
        elif arr[0]=="hgt":
            if arr[1][-2:]=="cm" and int(arr[1][:-2])>=150 and int(arr[1][:-2])<=193:
                self.hgt=True
            elif arr[1][-2:]=="in" and int(arr[1][:-2])>=59 and int(arr[1][:-2])<=76:
                self.hgt=True
        elif arr[0]=="hcl" and arr[1][0]=="#":
            for x in arr[1][1:]:
                if (x >='a' and x<='f') or (x >='0' and x<='9'):
                    self.hcl=True
                else:
                    return #could cause edge case maybe cosnider rewriting
        elif arr[0]=="ecl" and (arr[1]=="amb" or arr[1]=="blu" or arr[1]=="brn" or arr[1]=="gry" or arr[1]=="grn" or arr[1]=="hzl"or arr[1]=="oth"):
            self.ecl=True 
        elif arr[0]=="pid" and len(arr[1])==9:
            for x in arr[1]:
                if(x <'0' and x>'9'):
                    return
            self.pid=True
        elif arr[0]=="cid":
            self.cid=True   
        self.check()

def main():
    Processor=Checker()
    with open('day4.txt') as f:
        #hcl:#6b5442 ecl:brn iyr:2019
        for x in f:#takes a string line
            x=x.strip()#gets rid of /n at end of line
            array=re.split(' ',x)#supposed to split the string by spaces
            for i in array:
                Processor.checker(i)
                #print(array[i])
            #for x in range(0,(len(array) -1), 2):
            #    checker(array[x], array[x+1])
    print(Processor.counter)
if __name__ == "__main__":
    main()